### ISSUE TITLE
[Bug] 'repl' user password not overwritten in redeployment with undeleted PVCs 

### ISSUE THREAD

**Issue by philosapiens** [mariadb-operator/mariadb-operator#153]
**Describe the bug**

Deploying a replicated MariaDB, then deleting the deployment without deleting the PVCs leads to the following error and log message:

`[Warning] Access denied for user 'repl'@'<hostname>' (using password: YES)`

Once the MariaDB is redeployed, the operator checks for an existing secret. Since it was deleted when deleting the deployment, a new replication password is created. The operator then tries to create the user, but since it's already existing, it is not successful. The password of the already existing user 'repl' is not overwritten, leading to the log message above.

**Expected behaviour**

The replication password shall be overwritten with the current value stored within the secret.

**Steps to reproduce the bug**

1. Deploy a replicated MariaDB 
2. Delete MariaDB deployment without manually deleting the PVCs
3. Redeploy the same replicated MariaDB

**Additional context**

This might not be a day-to-day use case, but should be looked at.

**Environment details**:
- Openshift version: 4.10.59
- Kubernetes version: v1.23.17+16bcd69
- mariadb-operator version: v0.15.1
- Install method: helm
- Install flavour: recommended

**Comment by mmontes11** [comment]
Hey @philosapiens ! Thanks for bringing this up

I agree that it may not be something that happens on a daily basis, but still worth fixing.

The problem I see here is that the `Secret` containing the `repl` password is owned by `Mariadb` and therefore it is garbage collected by Kubernetes whenever the latter one is deleted. However, the PVCs never get deleted, so we Mariadb is still using refering to this `Secret` value internally.

What if the `Secret` wasn't owned by `Mariadb`? This way if we recreate the `Mariadb` we will be reusing the same `Secret`. I don't think it should be very hard to implement. Would you like to give it a try? Contributions are welcome!

Context:
https://github.com/mariadb-operator/mariadb-operator/blob/main/pkg/controller/replication/config.go#L218
https://github.com/mariadb-operator/mariadb-operator/blob/46fd744beb37eb952d9a580e319df79e8ca12447/pkg/controller/secret/controller.go#L45
https://github.com/mariadb-operator/mariadb-operator/blob/46fd744beb37eb952d9a580e319df79e8ca12447/pkg/builder/secret_builder.go#L33

Thanks 🙏🏻

**Comment by philosapiens** [comment]
Hey @mmontes11,

I want to correct myself in the context of automatic deployments; the potential redeployment happens more often.

I like that the resources are garbage collected. Otherwise, these secrets would have to be cleaned manually in case of no new redeployment. What about adding a new `replPasswordSecretKeyRef` referencing a new secret for the replication password? By doing that, the password would not have to be auto-generated and could fulfil potential unmet password requirements. It should also solve the problem of changing replication passwords.

Alternatively, updating the `repl` user's password within each pod could also be an option to keep the secret from being garbage collected.

What do you think about this?

**Comment by mmontes11** [comment]
Hey @philosapiens !

>  What about adding a new replPasswordSecretKeyRef referencing a new secret for the replication password?

This makes total sense to me, we could potentially introduce the `spec.replication.replPasswordSecretKeyRef` :
```yaml
apiVersion: mariadb.mmontes.io/v1alpha1
kind: MariaDB
metadata:
  name: mariadb-repl
spec:
...
  replication:
    ...
    replica:
      replPasswordSecretKeyRef:
         name: repl
         key: password
      waitPoint: AfterCommit
      gtid: CurrentPos
      connectionTimeout: 10s
      connectionRetries: 10
      syncTimeout: 10s
```
The field would be optional and the operator would generate a random password if the user does not provide the field.

### 

Regarding the implementation, we could check if `spec.replication.replPasswordSecretKeyRef` is non nil here and read it from the reference using `r.refResolver`. We would also avoid making the call to `r.secretReconciler.ReconcileRandomPassword` in this situation:

https://github.com/mariadb-operator/mariadb-operator/blob/46fd744beb37eb952d9a580e319df79e8ca12447/pkg/controller/replication/config.go#L218

Thoughts?

**Comment by philosapiens** [comment]
Yes, that's exactly what I had in mind.

Making the field optional would mean that if the optional `replPasswordSecretKeyRef` is not used, the problem with redeployment would remain. I don't know if that's wanted behaviour.

Making it a forced value would marginally increase the deployment effort but would solve the issue completely.

**Comment by mmontes11** [comment]
This has been fixed via https://github.com/mariadb-operator/mariadb-operator/pull/178. To be released in `v0.0.20`, stay tuned!
